fairlearn/grid_search.py

```diff
@@ -0,0 +1,142 @@
+        the optimisation problem. If this is set then num_Ls must be None. The result array will
+        be equal in length to this array
+        :type lagrangian_multipliers: List of real numbers
+        
+        :param number_lagrangian_multipliers: Specifies the number of Lagrangian multipliers to
+        use in the optimisation problem. If this is set then lagrangian_multipliers must be None.
+        The result array will have as many entries as specified here
+
+        :return: The models corresponding to each value of the Lagrangian multiplier tested
+        :rtype: List of dictionaries. Each dictionary has fields "lagrangian_multiplier" and "model." Each
+        model will correspond to the input parameter learner after calling 'fit' on it (a deep
+        copy is made). The user is responsible for converting these objects to an actual model,
+        if further processing is required.
+        """
+        # Must specify either Ls or num_Ls
+        if not (lagrangian_multipliers is None) ^ (number_lagrangian_multipliers is None):
+            raise RuntimeError("Must specify either lagrangian_multipliers or number_lagrangian_multipliers")
+
+        # Check that x, y and protected_attribute have the same number of rows
+        # Check that y and protected_attribute are both 1d vectors
+        if len(y.shape) != 1:
+            raise RuntimeError("Supplied y not 1d vector")
+
+        if len(protected_attribute.shape) != 1:
+            raise RuntimeError("Supplied protected_attribute not 1d vector")
+
+        if protected_attribute.shape[0] != y.shape[0]:
+            raise RuntimeError("Supplied protected_attribute and y not same length")
+
+        if x.shape[0] != y.shape[0]:
+            raise RuntimeError("Supplied x and y do not have same number of rows")
+
+        # Check that A only has values 0 and 1
+        unique_protected_attribute_values = np.unique(protected_attribute)
+        if not np.array_equal(unique_protected_attribute_values, [0, 1]):
+            raise RuntimeError("Supplied protected_attribute labels not 0 or 1")
+
+        # Compute p0 and p1
+        # This will also check that Y is binary with values
+        # 0 and 1
+        p0, p1 = self._generate_p0_p1(y)
+
+        # If not supplied, generate array of trial Lagrangian multipliers
+        if lagrangian_multipliers is None:
+            limit = 1
+            if p0/p1 > 1:
+                limit = p0/p1
+            lagrangian_multipliers = np.linspace(-2*limit, 2*limit, number_lagrangian_multipliers)
+
+        result = []
+        for current_multiplier in lagrangian_multipliers:
+            # Generate weights array
+            sample_weights = self._generate_weights(y, protected_attribute, current_multiplier, p1/p0)
+
+            # Generate Y'
+            f = lambda x: 1 if x > 0 else 0
+            re_labels = np.vectorize(f)(sample_weights)
+
+            # Run the learner
+            mylearner = copy.deepcopy(learner)
```

Comment:
I'm not a huge fan of mylearner, but I'm also struggling to come up with a better name.

Since you already used "current_" above, maybe that's an option?

Reply:
Done